src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -156,6 +222,673 @@ MultiRouterPlanCreate(Query *originalQuery, Query *query,
+			IsA(targetEntry->expr, Var))
+		{
+			targetPartitionColumnVar = (Var *) targetEntry->expr;
+			break;
+		}
+	}
+
+	/* we should have found target partition column */
+	Assert(targetPartitionColumnVar != NULL);
+
+	integer4GEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTGreaterEqualStrategyNumber);
+	integer4LEoperatorId = get_opfamily_member(INTEGER_BTREE_FAM_OID, INT4OID,
+											   INT4OID,
+											   BTLessEqualStrategyNumber);
+
+	/* look up the type cache */
+	typeEntry = lookup_type_cache(targetPartitionColumnVar->vartype,
+								  TYPECACHE_HASH_PROC_FINFO);
+
+	/* probable never possible given that the tables are already hash partitioned */
+	if (!OidIsValid(typeEntry->hash_proc_finfo.fn_oid))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_FUNCTION),
+						errmsg("could not identify a hash function for type %s",
+							   format_type_be(targetPartitionColumnVar->vartype))));
+	}
+
+	/* generate hashfunc(partCol) expression */
+	hashFunctionExpr = makeNode(FuncExpr);
+	hashFunctionExpr->funcid = typeEntry->hash_proc_finfo.fn_oid;
+	hashFunctionExpr->args = list_make1(targetPartitionColumnVar);
+
+	/* hash functions always return INT4 */
+	hashFunctionExpr->funcresulttype = INT4OID;
+
+	/* generate hashfunc(partCol) >= shardMinValue OpExpr */
+	greaterThanAndEqualsBoundExpr =
+		(OpExpr *) make_opclause(integer4GEoperatorId,
+								 InvalidOid, false,
+								 (Expr *) hashFunctionExpr,
+								 (Expr *) MakeInt4Constant(shardInterval->minValue),
+								 targetPartitionColumnVar->varcollid,
+								 targetPartitionColumnVar->varcollid);
+
+	/* update the operators with correct operator numbers and function ids */
+	greaterThanAndEqualsBoundExpr->opfuncid =
+		get_opcode(greaterThanAndEqualsBoundExpr->opno);
+	greaterThanAndEqualsBoundExpr->opresulttype =
+		get_func_rettype(greaterThanAndEqualsBoundExpr->opfuncid);
+
+	/* generate hashfunc(partCol) <= shardMinValue OpExpr */
+	lessThanAndEqualsBoundExpr =
+		(OpExpr *) make_opclause(integer4LEoperatorId,
+								 InvalidOid, false,
+								 (Expr *) hashFunctionExpr,
+								 (Expr *) MakeInt4Constant(shardInterval->maxValue),
+								 targetPartitionColumnVar->varcollid,
+								 targetPartitionColumnVar->varcollid);
+
+	/* update the operators with correct operator numbers and function ids */
+	lessThanAndEqualsBoundExpr->opfuncid = get_opcode(lessThanAndEqualsBoundExpr->opno);
+	lessThanAndEqualsBoundExpr->opresulttype =
+		get_func_rettype(lessThanAndEqualsBoundExpr->opfuncid);
+
+	/* finally add the quals */
+	if (subqery->jointree->quals == NULL)
+	{
+		subqery->jointree->quals = (Node *) greaterThanAndEqualsBoundExpr;
+	}
+	else
+	{
+		subqery->jointree->quals = make_and_qual(subqery->jointree->quals,
+												 (Node *) greaterThanAndEqualsBoundExpr);
+	}
+
+	subqery->jointree->quals = make_and_qual(subqery->jointree->quals,
+											 (Node *) lessThanAndEqualsBoundExpr);
```

Comment:
Personally I'd find it a bit easier to understand if you anded the two bound condition, and then did the quals == null check et al.